rig/place_and_route/place/sa/c_kernel.py

```
@@ -0,0 +1,129 @@
+"""Wraps the "rig_c_sa" Python C extension library in a Kernel object."""
+
+from collections import defaultdict
+
+from itertools import chain, repeat
+
+from six import iteritems
+
+import pkg_resources
+
+# An optional Rig dependency.
+from rig_c_sa import ffi
+import rig_c_sa
+
+
+# Make sure installed rig_c_sa version is compatible.
+pkg_resources.require("rig_c_sa>=0.3.1,<1.0.0")
+
+
+class CKernel(object):  # pragma: no cover
+    """An implementation of the Simulated Annealing placement algorithm kernel
+    written in C.
+
+    Thanks to being written in C, this kernel is 50-150x faster than
+    :py:class:`~rig.place_and_route.place.sa.python_kernel.PythonKernel` while
+    maintaining the same placement quality. Since installing the C
+    kernel requires a compiler it is not included in the standard Rig package.
+    The C kernel can be installed using::
+
+        $ pip install rig_c_sa
+
+    The `rig_c_sa <https://github.com/project-rig/rig_c_sa>`_ package is a
+    Python/C library containing the C components of the algorithm. Installation
+    requires that you have C compiler installed along with `libffi
+    <https://sourceware.org/libffi/>`_ which are readily available in the
+    package managers of real operating systems.
+    """
+
+    def __init__(self, vertices_resources, movable_vertices, fixed_vertices,
+                 initial_placements, nets, machine, random, **kwargs):
+
+        # Seed the random number generator
+        rig_c_sa.srand(random.getrandbits(32))
+
+        # Allocate the basic state for the kernel
+        self.s = rig_c_sa.sa_new(machine.width, machine.height,
+                                 len(machine.chip_resources),
+                                 len(vertices_resources),
+                                 len(nets))
+        assert self.s != ffi.NULL
```

If this is likely to occur and has a known fix this should raise an error.